dd-trace-propagation/src/tracecontext.rs

```diff
@@ -84,85 +84,117 @@ fn inject_traceparent(context: &SpanContext, carrier: &mut dyn Injector) {
+    if let Some(origin) = context.origin.as_ref() {
+        let origin_encoded = encode_tag_value(
             TRACESTATE_ORIGIN_FILTER_REGEX.replace_all(origin.as_ref(), INVALID_CHAR_REPLACEMENT),
-        )
-    }) {
-        tracestate_parts.push(format!("{TRACESTATE_ORIGIN_KEY}:{origin}"));
-    };
+        );
 
-    let last_parent_id = if context.is_remote {
-        match context.tags.get(DATADOG_LAST_PARENT_ID_KEY) {
-            Some(id) => id.to_string(),
-            None => format!("{:016x}", context.span_id), // TODO: is this correct?
+        if dd_parts.len()
+            + TRACESTATE_DD_PAIR_SEPARATOR.len()
+            + TRACESTATE_ORIGIN_KEY.len()
+            + 1
+            + origin_encoded.len()
+            < TRACESTATE_DD_KEY_MAX_LENGTH
+        {
+            dd_parts.push_str(TRACESTATE_DD_PAIR_SEPARATOR);
+            dd_parts.push_str(TRACESTATE_ORIGIN_KEY);
+            dd_parts.push(':');
+            dd_parts.push_str(&origin_encoded);
         }
-    } else {
-        format!("{:016x}", context.span_id)
-    };
-
-    tracestate_parts.push(format!("{TRACESTATE_LAST_PARENT_KEY}:{last_parent_id}"));
-
-    let tags = context
-        .tags
-        .keys()
-        .filter(|key| key.starts_with(DATADOG_PROPAGATION_TAG_PREFIX))
-        .map(|key| {
-            let t_key = format!(
-                "{TRACESTATE_DATADOG_PROPAGATION_TAG_PREFIX}{}",
-                TRACESTATE_TAG_KEY_FILTER_REGEX.replace_all(&key[6..], INVALID_CHAR_REPLACEMENT)
-            );
-
-            let value = encode_tag_value(
-                TRACESTATE_TAG_VALUE_FILTER_REGEX
-                    .replace_all(&context.tags[key], INVALID_CHAR_REPLACEMENT),
-            );
-
-            format!("{t_key}:{value}")
-        })
-        .collect::<Vec<String>>()
-        .join(TRACESTATE_DD_PAIR_SEPARATOR);
-
-    if !tags.is_empty() {
-        tracestate_parts.push(tags);
     }
 
-    let dd = tracestate_parts
-        .into_iter()
-        .reduce(|dd, part| {
-            if dd.len() + part.len() + 1 < TRACESTATE_DD_KEY_MAX_LENGTH {
-                format!("{dd}{TRACESTATE_DD_PAIR_SEPARATOR}{part}")
+    // Build last parent id part
+    let last_parent_id_part_start =
+        dd_parts.len() + TRACESTATE_DD_PAIR_SEPARATOR.len() + TRACESTATE_LAST_PARENT_KEY.len() + 1;
+    if last_parent_id_part_start + 16 < TRACESTATE_DD_KEY_MAX_LENGTH {
+        // 16 chars for hex span_id
+        dd_parts.push_str(TRACESTATE_DD_PAIR_SEPARATOR);
+        dd_parts.push_str(TRACESTATE_LAST_PARENT_KEY);
+        dd_parts.push(':');
+
+        if context.is_remote {
+            if let Some(id) = context.tags.get(DATADOG_LAST_PARENT_ID_KEY) {
+                dd_parts.push_str(id);
             } else {
-                dd
+                dd_parts.push_str(&format!("{:016x}", context.span_id));
             }
-        })
-        .unwrap_or_default();
+        } else {
+            dd_parts.push_str(&format!("{:016x}", context.span_id));
```

Comment:
Could be `let _ = write!(&mut dd_parts, "{:016x}", context.span_id);` to save one allocation
(the `let _ =` is necessary because write! returns a Result, although we're sure it's going to be Ok)